doc/dev/rule-streaming.md

```diff
@@ -0,0 +1,241 @@
+
+This is in fact how Dune evaluates globs during the build. Indeed,
+when writing `dune` files we work in an imaginary world where both the
+source files and the generated files are present. So when we write
+`(deps (glob_files *.txt))`, this `*.txt` denotes both `.txt` files
+that are present on disk in the source tree but also as the ones that
+can be generated by the build.
+
+In practice, to evaluate `(glob_files *.txt)` in directory `d`, Dune
+calls `Load_dir.load_dir ~dir:d` and filter the list of files that can
+be built. Similarly, when Dune needs to build a file
+`_build/default/src/x`, it first calls `Load_dir.load_dir` with
+`_build/default/src` and then looks up for a rule that has `x` has
+target in the returned `Loaded.t`. The `Load_dir.load_dir` is
+memoised, so it can be called multiple times during the build without
+guilt.
+
+While `Load_dir` is responsible for driving the production of rules,
+it is part of `dune_engine` which doesn't know about `dune` files and
+doesn't know about OCaml libraries or OCaml compilation in general. So
+it is not responsible for actually producing the build rules that
+allow to build Dune projects. Instead, `Load_dir` defers the actual
+production of rules to a callback that it obtains via
+`Build_config`. Inside Dune, this callback is implemented by the
+`Gen_rules` module inside the `dune_rules` library. `dune_rules` is
+the library that is responsible for parsing, interpreting and
+compiling `dune` files down to low-level build rules.
+
+### `Dune_rules.Gen_rules`
+
+The entry of `Dune_rules.Gen_rules` is the `gen_rules` function. It's
+API looks like:
+
+```ocaml
+val gen_rules : dir:Path.Build.t -> Dune_engine.Rules.t Memo.Build.t
+```
+
+Where `Dune_rules.Rules.t` is a "raw" set of rules. Raw in the sense
+that there is no overlap checks or any other checks. During the rule
+production phase, we merely accumulate a set of rules that is later
+processed. The API of `gen_rules` is in fact a bit more complex, but
+the above definition is enough for the purpose of this document.
+
+The first thing `gen_rules` do is analyse the directory it is
+given. If the directory correspond to a source directory with a `dune`
+file, `gen_rules` will dispatch the call to the part of `dune_rules`
+that parses and interprets the `dune` file. This is the simplest case,
+but even in this case there are some things worth mentioning.
+
+For instance, when compiling an OCaml library dune stores the
+artifacts for the library in generated dot-directories. For instance,
+the cmi files for library `foo` living in source directory `src` will
+end up in `_build/default/src/.foo.objs/byte`. We could produce these
+rules when `gen_rules` is called with direcotry
+`_build/default/src/.foo.objs/byte`, however that would spread out the
+logic for interpreting `library` stanzas. It is much simpler to
+produce all the build rules corresponding to a `library` stanza in one
```

Comment:
> It is much simpler to produce all the build rules corresponding to a `library` stanza in one go.

Reading about all the complexity that this leads to, I'm not convinced that it's the optimal design. Maybe splitting up rules for `lilbrary` is in fact going to be simpler.

Reply:
I tried splitting the library rules this way a few times with poor results. This problem also prevents us from adding features like targets for inferred mli's. In general, it really is an annoying limitation that forces to "invert" the logic of how we want to write our rules.